### ISSUE TITLE
D3 issue

### ISSUE THREAD

**Issue by babycicak** [anselal/antminer-monitor#21]
D3 really a lot problem, it will be great if can apply this code in to system.

https://gist.github.com/tperalta82/6e11253cd4b9cf9c5c6fa15bbf046d4f

; D

**Comment by gnanet** [comment]
I am using a check code in php on a local webserver, which get checked by url with m:monit and a php code which does an ssh restart of cgminer.sh if too much X-s are there

**Comment by babycicak** [comment]
that's great of you! by having xxx is the D3 fault, poor programming. I hope this auto restart will be add in new features of this software soon.

**Comment by anselal** [comment]
After a conversation with @babycicak he pointed me to this thread https://bitcointalk.org/index.php?topic=2198580.20
A user (`seizu`) claims that the errors are caused because bitmain overclocked the D3s. The first batch came with 15GH/s and other batches followed with 17GH/s and 19GH/s. When he underclocked the D3 to its original frequency at `437` all the errors where gone.
Another thing I talked about with @babycicak is that he was using niceahash as a pool which in my opinion isn't that good. I tried it once and I had many connection problems. Though unrelated with the chip errors I suggested him to try another pool and let us now if the error was fixed.
The reboot feature is a nice one but I am sceptical about the auto-reboot, since if the miner is damaged for some reason and it always reboots 1) that won't fix the issue 2) you still lose money while rebooting.
The best think we can do is try to find what caused the problem in the first place and only after that we can figure out a solution to the problem.

**Comment by gnanet** [comment]
> The reboot feature is a nice one but I am sceptical about the auto-reboot, since if the miner is damaged for some reason and it always reboots 1) that won't fix the issue 2) you still lose money while rebooting.

following applies mostly to the procedures prior knowing that a default-clocking may help:
For this part we use [monit](https://mmonit.com/monit/) where we can define 
`if 6 restarts within 18 cycles then unmonitor`
Where **cycle** means 300 secs, **restart** means only restart the cgminer (over ssh, or the api if it allows this request), and we send an alert on unmonitor which means we have to check manually the state of that miner

**Comment by anselal** [comment]
@gnanet I am aware of monit but it runs only on POSIX systems (Linux/UNIX). We have many Windows users running this app. 
If we find a good solution for this problem I can add it to the app.
Currently I am prompting users to downgrade their D3 and I am awaiting feedback to see if it actual fixes those errors.
So far one user downgraded and gave positive feedback.

**Comment by anselal** [comment]
Bitmain released a new firmware for the Antminer D3 which fixes the issue that some miner's chip status is 'x' after running for several hours or days.
https://shop.bitmain.com/support.htm?pid=00720170915192851224mJ36J1ap066E
I am closing this issue since the reboot feature is requested in #1
